internal/pool/pool.go

```diff
@@ -53,18 +62,23 @@ type ConnPool struct {
 	poolTimeout time.Duration
 	idleTimeout time.Duration
 
-	conns     *connList
-	freeConns *connStack
-	stats     PoolStats
+	queue chan struct{}
+
+	connsMu sync.Mutex
+	conns   []*Conn
+
+	freeConnsMu sync.Mutex
+	freeConns   []*Conn
 
-	_closed int32
+	stats PoolStats
 
+	_closed int32 // atomic
 	lastErr atomic.Value
 }
 
 var _ Pooler = (*ConnPool)(nil)
 
-func NewConnPool(dial dialer, poolSize int, poolTimeout, idleTimeout time.Duration) *ConnPool {
+func NewConnPool(dial dialer, poolSize int, poolTimeout, idleTimeout, idleCheckFrequency time.Duration) *ConnPool {
```

Comment:
It is a bit complicated since sometimes I hold lock to do 2 actions to max performance, e.g.:

```
	p.connsMu.Lock()
	if cn != nil {
		p.remove(cn, errConnStale)
	}
	p.conns = append(p.conns, newcn)
	p.connsMu.Unlock()
```